src/test/regress/upgrade/config.py

```diff
@@ -1,68 +1,232 @@
+REGULAR_USER_NAME = "regularuser"
+SUPER_USER_NAME = "postgres"
 
-MASTER = 'master'
+CUSTOM_TEST_NAMES = ["custom_sql_test", "custom_create_test"]
+
+BEFORE_PG_UPGRADE_SCHEDULE = "./before_pg_upgrade_schedule"
+AFTER_PG_UPGRADE_SCHEDULE = "./after_pg_upgrade_schedule"
+
+CUSTOM_CREATE_SCHEDULE = "./custom_create_schedule"
+CUSTOM_SQL_SCHEDULE = "./custom_sql_schedule"
+
+AFTER_CITUS_UPGRADE_COORD_SCHEDULE = "./after_citus_upgrade_coord_schedule"
+BEFORE_CITUS_UPGRADE_COORD_SCHEDULE = "./before_citus_upgrade_coord_schedule"
+MIXED_BEFORE_CITUS_UPGRADE_SCHEDULE = "./mixed_before_citus_upgrade_schedule"
+MIXED_AFTER_CITUS_UPGRADE_SCHEDULE = "./mixed_after_citus_upgrade_schedule"
+
+CITUS_CUSTOM_TEST_DIR = "./tmp_citus_test"
+
+MASTER = "master"
 # This should be updated when citus version changes
-MASTER_VERSION = '10.2'
+MASTER_VERSION = "10.2"
 
 HOME = expanduser("~")
 
 
 CITUS_VERSION_SQL = "SELECT extversion FROM pg_extension WHERE extname = 'citus';"
 
 
-class CitusUpgradeConfig():
+def find_free_port():
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind(("", 0))
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return s.getsockname()[1]
+
+
+class NewInitCaller(type):
+    def __call__(cls, *args, **kwargs):
+        obj = type.__call__(cls, *args, **kwargs)
+        obj.init()
+        return obj
+
+
+class CitusBaseClusterConfig(object, metaclass=NewInitCaller):
+
+    data_dir_counter = 0
 
     def __init__(self, arguments):
-        self.bindir = arguments['--bindir']
-        self.pre_tar_path = arguments['--citus-pre-tar']
-        self.post_tar_path = arguments['--citus-post-tar']
-        self.pg_srcdir = arguments['--pgxsdir']
-        self.temp_dir = './tmp_citus_upgrade'
-        self.datadir = self.temp_dir + '/data'
+        if "--bindir" in arguments:
+            self.bindir = arguments["--bindir"]
+        self.pg_srcdir = arguments["--pgxsdir"]
+        self.temp_dir = CITUS_CUSTOM_TEST_DIR
+        self.worker_amount = 2
+        self.user = REGULAR_USER_NAME
+        self.is_mx = False
         self.settings = {
-            'shared_preload_libraries': 'citus',
-            'citus.node_conninfo': 'sslmode=prefer',
-            'citus.enable_version_checks' : 'false'
+            "shared_preload_libraries": "citus",
+            "citus.node_conninfo": "sslmode=prefer",
         }
-        self.mixed_mode = arguments['--mixed']
+        self.add_coordinator_to_metadata = False
+
+    def init(self):
+        self._init_node_name_ports()
+
+        self.datadir = self.temp_dir + "/data"
+        self.datadir += str(CitusBaseClusterConfig.data_dir_counter)
+        self.input_dir = self.datadir
+        self.output_dir = self.datadir
+        self.output_file = os.path.join(self.datadir, "run.out")
+        if self.worker_amount > 0:
+            self.chosen_random_worker_port = self.random_worker_port()
```

Comment:
I think some logs on the terminal about which worker is picked + which config we are running would be beneficial

Reply:
It makes sense but currently with parallelism (even without parallelism) the output on the terminal is quite messy because the messages are written to the same terminal concurrently. Maybe we should think about this a bit